Update the sample controller

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug

/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
This PR updates the sample controller to track developments of the past few years and rationalize some names a little.

This PR adds the specification of FieldManager in Options structs, for playing more nicely with server-side apply.

This PR switches from putting encoded strings to decoded structs in the workqueue, removing a place where parsing and error handling must be done.

This PR replaces `obj interface{}` with `item interface{}` where dealing with workqueue contents before type discrimination, because (a) it never was an API object anyway, and (b) the term "item" is used in much of the English description of what is in the workqueue.

This PR replaces `key` (a `string`) with `objectRef` (a `cache.ObjectName`) because of the type change and to be more clear about the specific meaning.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Brush up the sample controller

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

dims commented:
/verify-cla

dims commented:
/check-cla

dims commented:
/easycla

munnerz commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -342,21 +338,20 @@ func (c *Controller) updateFooStatus(foo *samplev1alpha1.Foo, deployment *appsv1
+	if objectRef, err := cache.ObjectToName(obj); err != nil {
 		utilruntime.HandleError(err)
 		return
+	} else {
```
we don't need an `else` block here as the `err != nil` block above already `returns`

MikeSpreitzer replied:
By using the `else` clause here the variable `objectRef` remains visible. The alternative is more verbose: declare a variable named `objectRef` with the right type, then an if statement that assigns to objectRef in the test, then the workqueue.Add.

dims replied:
✅

munnerz commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -61,6 +61,8 @@ const (
 	// is synced successfully
 	MessageResourceSynced = "Foo synced successfully"
+	// FieldManager distinguishes this controller from other things writing to API objects
+	FieldManager = "Foo-controller"
```
Should this be `foo-controller`? (aka all lower case?)

MikeSpreitzer replied:
Is this a documented norm or requirement? If so, where?

deads2k replied:
Not exactly documented, but our first 40 or so are all lower-case: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/names/controller_names.go#L44-L86

MikeSpreitzer replied:
Thanks for the pointer. I will change this one to "foo-controller".

dims replied:
✅

munnerz reviewed (commented):
thanks! these changes all make a lot of sense to me :) just a few comments

MikeSpreitzer commented:
What's going wrong with the CLA check? I have long been an IBM contributor to Kubernetes.

MikeSpreitzer commented:
@spzala ?

spzala commented:
/easycla

spzala commented:
> @spzala ?

@MikeSpreitzer seems like a blip with CLA bot. I also ran into a similar issue recently on a non-CNCF LF project. For this PR, I re-run the bot and the CLA check is green now. Thanks!

seans3 commented:
/triage accepted

deads2k commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -196,50 +198,50 @@ func (c *Controller) runWorker(ctx context.Context) {
 // processNextWorkItem will read a single work item off the workqueue and
 // attempt to process it, by calling the syncHandler.
 func (c *Controller) processNextWorkItem(ctx context.Context) bool {
```
This aligns to the example in our docs: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md .  there's been a little drift over time and we should collapse the discrepancy.

MikeSpreitzer replied:
@deads2k , I am not sure I understand your point. Which are you suggesting should change, and how?

deads2k commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -196,50 +198,50 @@ func (c *Controller) runWorker(ctx context.Context) {
 		// period.
-		defer c.workqueue.Done(obj)
-		var key string
+		defer c.workqueue.Done(item)
```
i wonder at what point we diverged. I'm not clear on why this is paired near the Get as in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md to ensure it's not leaked.

MikeSpreitzer replied:
Yeah, I think it would be better to do the `defer` as in controllers.md

deads2k commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -196,50 +198,50 @@ func (c *Controller) runWorker(ctx context.Context) {
 
 	// We wrap this block in a func so we can defer c.workqueue.Done.
-	err := func(obj interface{}) error {
+	err := func(item interface{}) error {
```
if we eliminated this function it would match our example and existing controllers

MikeSpreitzer replied:
Yeah. I assume somebody put it here for a reason.

MikeSpreitzer replied:
Git blame tells me that this func was here from the start --- which was PR #52753 from @munnerz .

deads2k replied:
Let's get this eliminated to match the long standing doc in https://github.com/kubernetes/community/pull/7741 that we use for kubernetes controllers.

MikeSpreitzer replied:
@deads2k: I have removed that inner func.

deads2k commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -280,15 +276,15 @@ func (c *Controller) syncHandler(ctx context.Context, key string) error {
 	// If the resource doesn't exist, we'll create it
 	if errors.IsNotFound(err) {
-		deployment, err = c.kubeclientset.AppsV1().Deployments(foo.Namespace).Create(context.TODO(), newDeployment(foo), metav1.CreateOptions{})
+		deployment, err = c.kubeclientset.AppsV1().Deployments(foo.Namespace).Create(context.TODO(), newDeployment(foo), metav1.CreateOptions{FieldManager: FieldManager})
```
Adding a field manager to non-apply requests is a little unusual.  What does this do for future us?

MikeSpreitzer replied:
I thought it was normal to supply the FieldManager in all requests. I see that it is in all the XXXOptions types, and the apiserver writes into the ManagedFields in all write operations, so those are consistent with supplying FieldManger in all writes.

dims replied:
@deads2k is this ok to leave in?

sttts replied:
Seems to be correct to pass the field manager here too. ManagedFields is always written.

MikeSpreitzer commented:
The force-push to d06acb40658 is a response to the review above. I changed the FieldManger value and moved the defer of Done to be as early as possible.

MikeSpreitzer commented:
See also https://github.com/kubernetes/community/pull/7741

dims commented:
@MikeSpreitzer please rebase!

dims commented:
/assign @sttts @jpbetz

dims commented:
/lgtm

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -196,64 +198,56 @@ func (c *Controller) runWorker(ctx context.Context) {
+	// there was a failure so be sure to report it.  This method allows for
+	// pluggable error handling which can be used for things like
+	// cluster-monitoring.
+	utilruntime.HandleError(fmt.Errorf("error syncing '%#v': %s, requeuing", objRef, err.Error()))
```
while we are at it: `HandleErrorWithContext`. It also supports structured logging with key-value pairs.

MikeSpreitzer replied:
done.

MikeSpreitzer commented:
The force-push to 7848612de8d applies the suggestion from @sttts above and does a squash.

sttts commented:
/lgtm
/approve

MikeSpreitzer commented:
BTW, there is also https://github.com/kubernetes/community/pull/7741

MikeSpreitzer commented:
/hold
so I can change all those other calls to HandleError

MikeSpreitzer committed:
Finish switching to utilruntime.HandleErrorWithContext

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer commented:
/remove-hold

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -349,12 +349,12 @@ func (c *Controller) handleObject(obj interface{}) {
 		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
 		if !ok {
-			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			utilruntime.HandleErrorWithContext(context.Background(), nil, "Error decoding object, invalid type", "type", fmt.Sprintf("%T", obj), "value", obj)
```
is it best practice to put a whole object into a (logging) key-value pair?

MikeSpreitzer replied:
I don't know about "best". It has been _my_ practice, on the grounds that more information is better for someone who has to debug the problem. Are you saying that there is a problem with this?

MikeSpreitzer replied:
The force-push to af7fb3cf4a3 adds a comment telling the developer to think about information disclosure risk here.

sttts replied:
This creates giant log lines potentially. Am not a fan.

@pohly are there any guidelines?

pohly replied:
I share the concern about information disclosure and log entry size. I'm not sure whether we have any such guidelines written down, though.

If it was some known type with known harmless fields, then dumping it may make sense. But for an unknown type I would not include the content of the object.

MikeSpreitzer replied:
The developer generally knows the type of object, so can usually make the call here.

The force-push to fdbf0bbb2d8 changes the code to omit the value and changes the comment to suggest adding the value based on size and sensitivity of content.

sttts commented:
/lgtm

jpbetz commented:
Excellent, thanks @MikeSpreitzer
